File: 20-easy-gemstones/gemstones.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}

/*
 * Complete the 'gemstones' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts STRING_ARRAY arr as parameter.
 */

function gemstones(arr) {
    // Write your code here
    let myMap = new Map();
    arr.forEach(s=>{
        let set = new Set(s.split(''));
        let arr = [...set];
        arr.forEach(e=>{
            if(myMap.has(e)){ 
                myMap.set(e, myMap.get(e) + 1); 
            }else{             
                myMap.set(e,1);
            }
            // myMap.set(e, myMap.has(e) ? myMap.get(e) + 1 : 1 );
        });
    });
    let gemcount = 0;
    myMap.forEach(v=>{if(v==arr.length) ++gemcount;});
    return gemcount;
}

function gemstones_alt(arr) {
    // Write your code here
    let myMap = new Map();
    arr.forEach( s =>{
        let set = new Set(s.split(''));
        [...set].forEach( e =>{
            myMap.set(e, myMap.has(e) ? myMap.get(e) + 1 : 1 );
        });
    });
    let gemcount = 0;
    myMap.forEach( v =>{if(v==arr.length) ++gemcount;});
    return gemcount;
}


function main() {
    
    const n = parseInt(readLine().trim(), 10);
    
    let arr = [];
    
    for (let i = 0; i < n; i++) {
        const arrItem = readLine();
        arr.push(arrItem);
    }
    
    const result = gemstones(arr);

    //const ws = fs.createWriteStream(process.env.OUTPUT_PATH);
    if(process.env.OUTPUT_PATH) {
        let ws = fs.createWriteStream(process.env.OUTPUT_PATH);
        ws.write(result + '\n');
        ws.end();
    }else{
        console.log(result);
    }
}

File: 20-easy-gemstones/gemstones.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

import java.util.HashMap;
import java.util.HashSet;

class Result {

    /*
     * Complete the 'gemstones' function below.
     *
     * The function is expected to return an INTEGER.
     * The function accepts STRING_ARRAY arr as parameter.
     */

    public static int gemstones(List<String> arr) {
        // Write your code here

        HashMap<Character, Integer> myMap = new HashMap<Character, Integer>();

        arr.forEach(e -> {
            char[] ca = e.toCharArray();
            HashSet<Character> mySet = new HashSet<Character>();
            for(int i=0; i<ca.length; ++i){
                mySet.add(ca[i]);
            }
            mySet.forEach(c->{
                if( myMap.containsKey(c)){
                    myMap.put(c, myMap.get(c) + 1 );
                }else{
                    myMap.put(c, 1 );
                }
            });
        });
        int gemcount = 0;
        for(Map.Entry<Character,Integer> entry : myMap.entrySet()){
            if( entry.getValue() == arr.size() ){
                ++gemcount;
            }
        }
        return gemcount;
    }

}

public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        
        int n = Integer.parseInt(bufferedReader.readLine().trim());
        List<String> arr = IntStream.range(0, n).mapToObj(i -> {
            try {
                return bufferedReader.readLine();
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        })
        .collect(toList());
        bufferedReader.close();
        
        int result = Result.gemstones(arr);
        
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));
        
        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();
        bufferedWriter.close();
    }
}
